File: lib/billing/cron-auth.ts
import { NextRequest, NextResponse } from "next/server";

/**
 * Verify that a request has the correct CRON_SECRET header.
 * Returns null on success, or a NextResponse error on failure.
 * If CRON_SECRET is not set (dev mode), always allows the request.
 */
export function verifyCronSecret(req: NextRequest): NextResponse | null {
  const secret = process.env.CRON_SECRET;
  if (!secret) return null; // Dev mode — no protection

  const provided =
    req.headers.get("x-cron-secret") ??
    req.headers.get("authorization")?.replace("Bearer ", "");

  if (provided !== secret) {
    return NextResponse.json({ error: "Unauthorized" }, { status: 401 });
  }

  return null;
}

File: rustbill/crates/rustbill-server/src/middleware/cron_auth.rs
//! Cron secret authentication for scheduled task endpoints.

use crate::app::SharedState;
use axum::{
    extract::State,
    http::{Request, StatusCode},
    middleware::Next,
    response::{IntoResponse, Response},
    Json,
};

/// Middleware that requires cron secret OR valid admin session.
pub async fn require_cron_or_admin(
    State(state): State<SharedState>,
    req: Request<axum::body::Body>,
    next: Next,
) -> Response {
    // Check cron secret
    let cron_secret = req
        .headers()
        .get("x-cron-secret")
        .or_else(|| req.headers().get("authorization"))
        .and_then(|v| v.to_str().ok())
        .map(|s| s.strip_prefix("Bearer ").unwrap_or(s).to_string());

    if let Some(ref expected) = state.config.auth.cron_secret {
        if let Some(ref provided) = cron_secret {
            if provided == expected {
                return next.run(req).await;
            }
        }
    } else {
        // No cron secret configured — allow in dev mode
        return next.run(req).await;
    }

    // Fall back to session auth check
    let token = req
        .headers()
        .get("cookie")
        .and_then(|v| v.to_str().ok())
        .and_then(|cookies| {
            cookies.split(';').find_map(|c| {
                let c = c.trim();
                c.strip_prefix("session=").map(String::from)
            })
        });

    if let Some(token) = token {
        if let Ok(Some(user)) = rustbill_core::auth::validate_session(&state.db, &token).await {
            if user.role == rustbill_core::db::models::UserRole::Admin {
                return next.run(req).await;
            }
        }
    }

    (
        StatusCode::UNAUTHORIZED,
        Json(serde_json::json!({ "error": "Unauthorized" })),
    )
        .into_response()
}
